<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN">
<html>
	<head>
		<title>RNAstructure Command Line Help: File Formats</title>
		<meta http-equiv="Content-Type" content="text/html;charset=utf-8">
		<link href="Stylesheet.css" rel="stylesheet" type="text/css">
	</head>
	<body>
		<div class="bodydiv">
			<table cellpadding="0" cellspacing="0" border="0" class="mainTable">
				<tr class="headerRow">
					<td class="headerBox"><img src="icon.jpg" alt="RNAstructure logo"></td>
					<td class="headerLabel"><h2>RNAstructure Command Line Help<br>File Formats</h2></td>
					<td class="headerBox2">
						<ul id="navigator">
							<li><h3 class="noMarginOrPad"><a href="index.html">Contents</a></h3></li>
							<li><h3 class="noMarginOrPad"><a href="AlphabeticalIndex.html">Index</a></h3></li>
						</ul>
					</td>
				</tr>
				<tr>
					<td class="mainContent" colspan="3">
						<div>
							<a name="Sequence" id="Sequence"></a>
							<h3 class="leftHeader">Sequence File Format</h3>
                        	<p>The native sequence format is the sequence file with a .SEQ extension. <b>Lowercase nucleotides will be forced single stranded in structure prediction.</b> The file must have the following format:</p>
                        	<ol>
								<li>Comment lines must be at the beginning of the file and must start with a semicolon. There needs to be at least one comment line in the file.</li>
								<li>Any number of comments are allowed, but each line before the title must start with a semicolon.</li>
								<li>After the comments, a title must be given and must be contained on one line.</li>
								<li>The line after the title must start the sequence, entered from 5' to 3'. The sequence can be on any number of lines and can contain U or T (T's are translated into U's for RNA). Spaces are allowed in the sequence. Bases should generally be in capital letters. <b>Lowercase letters can be used to indicate a base that should be single-stranded in the predicted structure.</b> &quot;XXXX&quot; can be used in the sequence to indicate that some bases have been left out of the prediction.</li>
								<li>Finally, the sequence must end in 1.</li>
							</ol>

							<fieldset>
								<legend>SAMPLE</legend>
								<pre><i>
; (first line of file) Comments must start with a semicolon.
;
; There can be any number of comments.
A title must immediately follow on the next line and be on one line.
AAA GCGG UUTGTT UTCUTaaTCTXXXXUCAGG1
								</i></pre>
							</fieldset>
						</div>

						<br><hr>

						<div>
							<a name="FASTA" id="FASTA"></a>
							<h3 class="leftHeader">FASTA (Sequence File) Format</h3>
							<p>Sequences can also be provided to RNAstructure in FASTA format. For FASTA, the first line, a title line, needs to start with &quot;&gt;&quot;. Subsequent lines should only contain sequence and whitespace, which is ignored. <b>Lowercase nucleotides will be forced single stranded in structure prediction.</b></p>

							<fieldset>
								<legend>SAMPLE</legend>
								<pre><i>
&gt;Title of Sequence
AAA GCGG UUTGTT UTCUTaaTCTXXXXUCAGG
UUA GCCG UUTGTT UTCUTaaTCTGGG
								</i></pre>
							</fieldset>
						</div>

						<br><hr>

						<div>
							<a name="CT" id="CT"></a>
							<h3>CT File Format</h3>
							<p>A CT (Connectivity Table) file contains secondary structure information for a sequence. These files are saved with a CT extension. When entering a structure to calculate the free energy, the following format must be followed.</p>
							<ol>
								<li>Start of first line: number of bases in the sequence</li>
								<li>End of first line: title of the structure</li>
								<li>
									Each of the following lines provides information about a given base in the sequence. Each base has its own line, with these elements in order:
									<ul>
										<li>Base number: index n</li>
										<li>Base (A, C, G, T, U, X)</li>
										<li>Index n-1</li>
										<li>Index n+1</li>
										<li>Number of the base to which n is paired. No pairing is indicated by 0 (zero).</li>
										<li>Natural numbering. RNAstructure ignores the actual value given in natural numbering, so it is easiest to repeat n here.</li>
									</ul>
								</li>
							</ol>

							<p>
								The CT file may hold multiple structures for a single sequence. This is done by repeating the format for each structure without any blank lines between structures.
								<br><br>
								The CT file format is such that files generated by RNAstructure are compatible with mfold/Unafold (available from Michael Zuker), and many other software packages.
							</p>
						</div>

						<hr>

						<div>
							<a name="Constraint" id="Constraint"></a>
							<h3 class="leftHeader">Constraint File Format</h3>
							<p>Folding constraints are saved in plain text with a CON extension. These can be hand edited. For multiple entries of a specific type of constraint, entries are each listed on a separate line. Note that all specifiers, followed by "-1" or "-1 -1", are expected by RNAstructure. For all specifiers that take two arguments, it is assumed that the first argument is the 5'nucleotide. Nucleotides positions are specified from the 5' end, where the first nucleotide in the sequence is in position 1.The file format is as follows:</p>

							<pre><i>
DS:
XA
-1
SS:
XB
-1
Mod:
XC
-1
Pairs:
XD1 XD2
-1 -1
FMN:
XE
-1
Forbids:
XF1 XF2
-1 -1
							</i></pre>

							<ul>
								<li>XA: Nucleotides that will be double-stranded</li>
								<li>XB: Nucleotides that will be single-stranded (unpaired)</li>
								<li>XC: Nucleotides accessible to chemical modification</li>
								<li>XD1, XD2: Forced base pairs</li>
								<li>XE: Nucleotides accessible to FMN cleavage (a U that must be in a GU pair)</li>
								<li>XF1, XF2: Prohibited base pairs</li>
							</ul>

							<fieldset>
								<legend>SAMPLE</legend>
								<pre><i>
DS:
15
25
76
-1
SS:
17
18
20
35
-1
Mod:
2
15
-1
Pairs:
16 26
-1 -1
FMN:
-1
Forbids:
15 27
-1 -1
								</i></pre>
							</fieldset>
						</div>

						<br><hr>

						<div>
							<a name="SHAPE" id="SHAPE"></a>
							<h3 class="leftHeader">SHAPE Data File Format</h3>
							<p>The file format for SHAPE reactivity comprises two columns. The first column is the nucleotide number, and the second is the reactivity.</p>
							<p>Nucleotides for which there is no SHAPE data can either be left out of the file, or the reactivity can be entered as less than -500. Columns are separated by any white space.</p>
							<p>Note that there is no header information. Nucleotides 1 through 10 have no reactivity information. Nucleotide 11 has a normalized SHAPE reactivity of 0.042816. Nucleotide 12 has a normalized SHAPE reactivity of 0, which is NOT the same as having no reactivity when using the pseudo-energy constraints.</p>
							<p>By default, RNAstructure looks for SHAPE data files to have the file extension SHAPE, but any plain text file can be read.</p>

							<fieldset>
								<legend>SAMPLE</legend>
								<pre><i>
9    -999
10   -999
11   0.042816
12   0
13   0.15027
14   0.16201
								</i></pre>
							</fieldset>
						</div>

						<br><hr>
	
						<div>
							<a name="List" id="List"></a>
							<h3 class="leftHeader">List File Format</h3>
							<p>List files have a LIS extension. This file contains any number of sequences of any length or nucleic acid, each on its own line.</p>

							<fieldset>
								<legend>SAMPLE</legend>
								<pre><i>
CUGAGCCAAG
GGGCUCAACG
GGCGUGAGAAAC
								</i></pre>
							</fieldset>
						</div>

						<br><hr>
	
						<div>
							<a name="Offset" id="Offset"></a>
							<h3 class="leftHeader">Offset File Format</h3>
							<p>Offset files are plain text. The files contain two colums: the nucleotide followed by the offset value in kcal.</p>

							<fieldset>
								<legend>SAMPLE</legend>
								<pre><i>
1 -0.336512
2 -1.4448
3 -1.74126
4 -3.82745
								</i></pre>
							</fieldset>
						</div>

						<br><hr>
	
						<div>
							<a name="Experimental" id="Experimental"></a>
							<h3 class="leftHeader">Experimental Pair Bonus File Format</h3>
							<p>Bonus files are plain text. They are formatted as an <em>n</em>x<em>n</em> matrix of bonus values, where <em>n</em> is the length of the sequence.</p>

							<fieldset>
								<legend>SAMPLE</legend>
								<pre><i>
0.0   1.0   0.0   1.0   0.0
0.0   1.0   0.0   1.0   0.0
0.0   1.0   0.0   1.0   0.0
0.0   1.0   0.0   1.0   0.0
0.0   1.0   0.0   1.0   0.0
								</i></pre>
							</fieldset>
						</div>

						<br><hr>
	
						<div>
							<a name="Alignment" id="Alignment"></a>
							<h3 class="leftHeader">Alignment File Format</h3>
							<p>Alignment files are plain text. They are formatted as a nucleotide in the first sequence immediately followed by the nucleotide in the second sequence it's aligned to, separated by a space. Only one alignment pair can be on each line, and the last line of the file must be "-1 -1".</p>

							<fieldset>
								<legend>SAMPLE</legend>
								<pre><i>
10 12
11 13
-1 -1
								</i></pre>
							</fieldset>
						</div>
                        
                        <br><hr>
                        
                        <div>
						  <a name="NMR" id="NMR"></a>
						  <h3 class="leftHeader">NMR File Format</h3>
						  <p>NMR file provides experimantal NRM constraints to NAPSS. </p>
                          
						  <fieldset>
							<legend>SAMPLE</legend>
							<pre><i>
666
66
665(+RAY)6
67
65(+YAR)7
665(-RAY)65(-RGY)5(+YGR)5(+YAY)6
57
							</i></pre>
						  </fieldset>
                        </div>
                        
                        <br><hr>
	
						<div>
						  <a name="DotBracket" id="DotBracket"></a>
						  <h3 class="leftHeader">Dot Bracket File Format</h3>
						  <p>Dot bracket files are plain text. The encode a sequence and secondary structure.  The first line starts with a ">" character and a file title follows.  The next line contains the sequence.  The final line contains "." (unpaired nucleotide), "(" (nucleotide that is 5' in a pair), and ")" (nucleotide that is 3' in a pair.  RNAstructure/examples/bmorivector.dot provides a sample file.</p>
                          
						  <fieldset>
							<legend>SAMPLE</legend>
							<pre><i>&gt;  A stem-loop structure<br>GGGCGAAUUGGGUACCGGGCCC<br>((((...((((...))))))))</i>
						  </pre>
						  </fieldset>
					    </div>
                        
                         <br><hr>
                        
                      <div>
				      <a name="FASTA Alignment Format" id="FASTA-Align"></a>
						  <h3 class="leftHeader">FASTA Alignment Format</h3>
						  <p>This is the expected format for sequence alignments used by <a href="Multifind.html">Multifind</a>.</p>
                          
					    <fieldset>
					    <legend>SAMPLE</legend>
							<pre><em>&gt;NZ_GG697986.1 44776 45325<br>TCGT-----------TCTTTCCCTTGAATCTCTATGATTAGAACACTATCGTCCAACTGG-------------------AAATGATAATTTAATAATGTACACTTTTTATTTTGTAAGAA<br>&gt;NC_002953.3 2279812 2280398<br>TTTTCGTCCCGTAG-TTCTTCCATTGAGCCTCTATGATTAGAACACAATCGTCCGGTTATCATACGGCCTCCGCAAGCTAAATGATAATTTAATAATGGACACTTTTGATTGTTTAAGCA<br>&gt;NC_017337.1 2316381 2316961<br>TCGT-----------TCTTTCCCTTGAATATCTATGATTAGAACACTATCGTGCGTTTATCGTCCAGCCTCCGCAAGCTAAATGACAATTTAATAATGTACACTTTTGATTGTGTAAGCA<br>&gt;NC_003923.1 2300707 2301293<br>TTTTCGTCCCGTAG-TTCTTCCATTGAGCCTCTATGATTAGAACACAATCGTCCGGTTATCATACGGCCTCCGCAAGCTAAATGATAATTTAATAATGGACACTTTTGATTGTTTAAGCA<br>&gt;NC_017338.1 2290333 2290748<br>TCGT-----------TCTTTCCCTTGAATATCTATGATTAGAACAC-----------------------------------------------TAATGTACACTTTTGATTGTGTAAACA<br>&gt;NZ_JH806555.1 46083 46649<br>TCGT-----------TCTTTCCCTTGAACCACTATGATTAGAACACAATCGTCTGGTTATCGTCCACCCTCCGCAAGCTAAATGACAATTTAATAATGTACACTTTTGATTGTGTAAACA<br>&gt;NC_002951.2 2282533 2283102<br>TTATCATCCGATAGCTCTTTCCCTTGAATATCTATTATTAGAACACTATCGTACGGT-----------CTCCGCAAGCTAAATGACAATTTAATAATGTACACTTTTGATTGTGTAAGCA<br>&gt;NC_017340.1 2295263 2295829<br>TCGT-----------TCTTTCCCTTGAACCACTATGATTAGAACACAATCGTCTGGTTATCGTCCACCCTCCGCAAGCTAAATGACAATTTAATAATGTACACTTTTGATTGTGTAAACA</em></pre>
						  </fieldset>
                        </div>
					</td>
				</tr>
				<tr>
				  <td class="footer" colspan="3">
					<p>Visit <a href="http://rna.urmc.rochester.edu/RNAstructure.html">The Mathews Lab RNAstructure Page</a> for updates and latest information.</p>
				  </td>
				</tr>
			</table>
		</div>
	</body>
</html>
